src/rust/src/orchard_ffi/incremental_sinsemilla_tree_ffi.rs

```diff
@@ -0,0 +1,287 @@
+
+use zcash_primitives::transaction::components::Amount;
+
+use crate::orchard_ffi::{error, CppStreamReader, CppStreamWriter, ReadCb, StreamObj, WriteCb};
+
+pub const MERKLE_DEPTH: u8 = 32;
+
+//
+// Operations on Merkle frontiers.
+//
+
+#[no_mangle]
+pub extern "C" fn orchard_merkle_frontier_empty(
+) -> *mut bridgetree::Frontier<OrchardIncrementalTreeDigest, MERKLE_DEPTH> {
+    let empty_tree = bridgetree::Frontier::<OrchardIncrementalTreeDigest, MERKLE_DEPTH>::new();
+    Box::into_raw(Box::new(empty_tree))
+}
+
+#[no_mangle]
+pub extern "C" fn orchard_merkle_frontier_clone(
+    tree: *const bridgetree::Frontier<OrchardIncrementalTreeDigest, MERKLE_DEPTH>,
+) -> *mut bridgetree::Frontier<OrchardIncrementalTreeDigest, MERKLE_DEPTH> {
+    unsafe { tree.as_ref() }
+        .map(|tree| Box::into_raw(Box::new(tree.clone())))
+        .unwrap_or(std::ptr::null_mut())
+}
+
+#[no_mangle]
+pub extern "C" fn orchard_merkle_frontier_free(
+    tree: *mut bridgetree::Frontier<OrchardIncrementalTreeDigest, MERKLE_DEPTH>,
+) {
+    if !tree.is_null() {
+        drop(unsafe { Box::from_raw(tree) });
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn orchard_merkle_frontier_parse(
+    stream: Option<StreamObj>,
+    read_cb: Option<ReadCb>,
+) -> *mut bridgetree::Frontier<OrchardIncrementalTreeDigest, MERKLE_DEPTH> {
+    let reader = CppStreamReader::from_raw_parts(stream, read_cb.unwrap());
+
+    match bincode::deserialize_from(reader) {
+        Ok(parsed) => Box::into_raw(Box::new(parsed)),
+        Err(e) => {
+            error!("Failed to parse Orchard bundle: {}", e);
+            ptr::null_mut()
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn orchard_merkle_frontier_serialize(
+    tree: *const bridgetree::Frontier<OrchardIncrementalTreeDigest, MERKLE_DEPTH>,
+    stream: Option<StreamObj>,
+    write_cb: Option<WriteCb>,
+) -> bool {
+    let tree = unsafe {
+        tree.as_ref()
+            .expect("Orchard tree pointer may not be null.")
+    };
+
+    let writer = CppStreamWriter::from_raw_parts(stream, write_cb.unwrap());
+    match bincode::serialize_into(writer, tree) {
+        Ok(()) => true,
+        Err(e) => {
+            error!("{}", e);
+            false
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn orchard_merkle_frontier_append_bundle(
+    tree: *mut bridgetree::Frontier<OrchardIncrementalTreeDigest, MERKLE_DEPTH>,
+    bundle: *const orchard::Bundle<Authorized, Amount>,
+) -> bool {
+    let tree = unsafe {
+        tree.as_mut()
+            .expect("Orchard note commitment tree pointer may not be null.")
+    };
+    if let Some(bundle) = unsafe { bundle.as_ref() } {
+        for action in bundle.actions().iter() {
+            if !tree.append(&OrchardIncrementalTreeDigest::from_cmx(action.cmx())) {
+                error!("Orchard note commitment tree is full.");
+                return false;
+            }
+        }
+
+        true
+    } else {
+        false
```

Comment:
If a transaction contains no Orchard component, we should be able to successfully append it to the tree (as a no-op):
```suggestion
        true
```